### PR TITLE
feat: (flux) add flux bundle

### PR DESCRIPTION
**Pull request by tyzbit**
Closes https://github.com/kairos-io/kairos/issues/638

Tested and working on k3s with [this config](https://github.com/tyzbit/kairos-config/blob/e0d787f553240ad517c7be8630d08a218d137be1/k3s/auroraboot-config.yaml).

It creates a ConfigMap to ensure only one node tries to boot Flux at once.

VSCode had some things to say about formatting for `README.md`, I can remove if desired.

### CODE REVIEW

File: fluxcd/bootstrap.sh [hunk 1]
```diff
@@ -0,0 +1,104 @@
+#!/bin/bash
+
+set -x
+# Override with the `env` key if necessary
+export KUBECONFIG=/etc/rancher/k3s/k3s.yaml
```
Comment (jimmykarily) [inline fluxcd/bootstrap.sh:5]: This makes will only work on k3s. There may be kairos' out there that run different providers with different k8s distributions. On the other hand, I'm not sure how it could be set outside the bundle. In any case, maybe it's better to set it to `k3s.yaml` if not already set.

File: README.md [hunk 2]
```diff
@@ -108,14 +106,46 @@ The bundle does add a `certManager` block, that allow to change the version (cur
...
+This installs Flux and supports bootstrapping the cluster with [FluxCD](https://fluxcd.io/flux/cmd/flux_bootstrap/). It has logic so only one node will do the bootstrapping,
+by creating a ConfigMap in the `default` namespace named `flux-bootstrap`.
+By default, it will time out after trying for 30 minutes.
+
+```yaml
+#cloud-config
+
+bundles:
+  - targets:
+      # Installs to /usr/local/lib/extensions/flux/
```
Comment (jimmykarily) [inline README.md:128]: This is not true. One would have to specify the path (e.g. [like this](https://github.com/kairos-io/kairos/blob/750b0d1d00b029cea57e5913357438b80a85b2c3/tests/assets/live-overlay.tmpl#L9)) and not all paths are available at bundle installation time (I tried various to discover that :D ). 

`run://` bundles on the other hand, they get extracted to a temp directory and the `run.sh` script is called if one is there. This bundle needs to run after kubernetes starts and according to the docs this is how to set it up in the kairos config. I think this needs some playing with `rootfs_path` to get it to extract properly. I see [other bundles installing binaries](https://github.com/kairos-io/community-bundles/blob/4bb2da2de3311023d7ed8c4023fff221c46d9727/coco/run.sh#L33) so that path should be writable at installation time.

To summarize, I think this bundle should:

- install the flux binary using a `run.sh` script (in `/usr/local/bin` for example)
- install a bootstap.sh script in some writable directory using `run.sh` script (not sure what paths are available at that point, I would check what other bundles do).


Then in the kairos config, the bootstrap.sh script should be invoked at the right stage like it happens currently in provided README in this PR. This will ensure the boostrap.sh script will only run when Kubernetes has started.

It needs some trial and error to get this right. Let me know if you need help figuring out a test workflow.

File: README.md [hunk 3]
```diff
@@ -108,14 +106,46 @@ The bundle does add a `certManager` block, that allow to change the version (cur
...
+      - container://quay.io/kairos/community-bundles:flux_latest
+
+stages:
+  boot:
+    - name: "Bootstrap with Flux"
+      commands:
+        - bash /usr/local/lib/extensions/flux/bootstrap.sh &
```
[the remark this answers is not in the document]
Comment (tyzbit) [reply README.md:135]: If it ran in a run.sh script, it blocked booting in my testing because the script blocks until it either bootstraps or confirms that instance won't be doing the bootstrapping.

The root problem is flux needs to be bootstrapped after the cluster has started and I couldn't figure out where to define running it immediately after the cluster is bootstrapped. I'd love for it to match the `run://` convention with no extra steps.

File: fluxcd/bootstrap.sh (hunk 1 above)
Comment (tyzbit) [reply fluxcd/bootstrap.sh:5]: The environment variables for Flux can be defined in the cloud-config, that line just sets a default. Here's how:

```yaml
flux:
  env:
    KUBECONFIG: /home/user/.kube/config
```

This part reads the keys under `env` and sets all variables and overrides the default if a `KUBECONFIG` is set. https://github.com/tyzbit/community-bundles/blob/f66dc04ce10a9826c7e970136f4b8271f8ce5565/fluxcd/bootstrap.sh#L40-L50

I've added an example how to override the KUBECONFIG for non-k3s distros.

File: flux/flux-bootstrap.sh [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+active="false"
+
+if [[ "${#cmdline[@]}" -eq 0 ]]; then
+  info "Flux was not configured in cloud-config, not bootstrapping"
+  exit 0
+else
+  while [[ $retry_attempt -le $total_attempts ]]; do
+    if [[ "$active" != "true" ]]; then
+      # Ensure only one host tries to bootstrap, whichever makes the configmap first
+      if ! timeout $short kubectl version &> /dev/null; then
```
Comment (jimmykarily) [inline flux/flux-bootstrap.sh:88]:
```localhost:/home/kairos # kubectl version && echo "ok"
Client Version: v1.28.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.25.11+k3s1
WARNING: version difference between client (1.28) and server (1.25) exceeds the supported minor version skew of +/-1
ok
```

but 

```
localhost:/home/kairos # journalctl -f -u flux-bootstrap
Oct 19 10:56:51 localhost flux-bootstrap.sh[1738]: INFO:  Kubernetes API not ready yet, sleeping
Oct 19 10:56:51 localhost flux-bootstrap.sh[1738]: WARN: Install attempt 19 (of 120) failed, retrying in 15 seconds
Oct 19 10:57:06 localhost flux-bootstrap.sh[1738]: INFO:  Kubernetes API not ready yet, sleeping
Oct 19 10:57:06 localhost flux-bootstrap.sh[1738]: WARN: Install attempt 20 (of 120) failed, retrying in 15 seconds
Oct 19 10:57:21 localhost flux-bootstrap.sh[1738]: INFO:  Kubernetes API not ready yet, sleeping
Oct 19 10:57:21 localhost flux-bootstrap.sh[1738]: WARN: Install attempt 21 (of 120) failed, retrying in 15 seconds
Oct 19 10:57:36 localhost flux-bootstrap.sh[1738]: INFO:  Kubernetes API not ready yet, sleeping
Oct 19 10:57:36 localhost flux-bootstrap.sh[1738]: WARN: Install attempt 22 (of 120) failed, retrying in 15 seconds
Oct 19 10:57:51 localhost flux-bootstrap.sh[1738]: INFO:  Kubernetes API not ready yet, sleeping
Oct 19 10:57:51 localhost flux-bootstrap.sh[1738]: WARN: Install attempt 23 (of 120) failed, retrying in 15 seconds
```

I'm not sure what's going on. I will debug more.

Comment (jimmykarily) [reply flux/flux-bootstrap.sh:88]: I added these lines right before the check:

```
      echo $KUBECONFIG
      whoami
      kubectl version

```

and this is the output:

```
localhost:/home/kairos # /usr/local/bin/flux-bootstrap.sh
/home/csagan/.kube/config
root
Client Version: v1.28.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
The connection to the server localhost:8080 was refused - did you specify the right host or port?

```

Comment (jimmykarily) [reply flux/flux-bootstrap.sh:88]: something changes the value of KUBECONFIG :thinking:

Comment (jimmykarily) [reply flux/flux-bootstrap.sh:88]: I'm stupid :facepalm: . I copied and pasted the config you provided and that sets it. Ignore the noise.

File: README.md (hunk 2 above)
Comment (jimmykarily) [reply README.md]: Let's leave it as it is now with k3s hardcoded. If someone requests the bundle to work with other k8s distros, we'll figure it out then.

**Review by jimmykarily — APPROVED** [review]
Awesome work @tyzbit !

**Comment by mauromorales** [comment]
Fantastic job guys, thanks!

File: README.md [hunk 5]
```diff
@@ -108,14 +106,41 @@ The bundle does add a `certManager` block, that allow to change the version (cur
...
 ```
 
+### Flux
+
+This installs [FluxCD](https://fluxcd.io/flux/cmd/flux_bootstrap/) and supports
+automatically bootstrapping the cluster. Only one node will do the bootstrap.
+It will time out after trying for 30 minutes and it requires `systemd`.
+
+```yaml
+#cloud-config
```
[the remark this answers is not in the document]
Comment (tyzbit) [reply README.md:124]: Good call! Added.
